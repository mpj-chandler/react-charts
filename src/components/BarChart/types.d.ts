import Placement from '../../enums/Placement';

export interface SeriesDataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullSeriesDataPoint extends SeriesDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    points: SeriesDataPoint[];
}

export interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
}
