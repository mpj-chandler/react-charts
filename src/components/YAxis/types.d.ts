import { AxisConfig, SeriesData } from "../BarChart/types";

export interface AxisRange {
    min: number;
    max: number;
}

export interface YAxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}
