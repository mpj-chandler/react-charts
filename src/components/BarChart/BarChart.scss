.BarChart {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-areas: "title title" "y-axis plot" "blank x-axis";
  grid-template-columns: 20% 60% 20%;
  grid-template-rows: 10% 70% 20%;

  &__title {
    grid-area: title;
  }

  &__blank {
    grid-area: blank;
    z-index: 0;
  }
}
