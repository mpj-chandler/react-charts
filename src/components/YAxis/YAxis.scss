.YAxis {
  grid-area: y-axis;

  &__Svg {
    width: 100%;
    height: 100%;
  }

  &__Ticks {
    width: 100%;
    height: 100%;
  }

  &__TickLabels {
    font-size: calc(8px + 0.5vh);
  }
}
