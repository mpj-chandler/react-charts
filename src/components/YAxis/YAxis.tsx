import React from 'react';
import { AxisRange, YAxisTickProps } from './types';
import styles from './YAxis.scss';
import { SeriesData, SeriesDataPoint, Padding, AxisProps } from '../BarChart/types';
import { getYAxisRange } from '../../utils/getYAxisRange';

const XAxis: React.FC<AxisProps> = (props) => {
    const renderTicks: (props: YAxisTickProps) => JSX.Element = (props: YAxisTickProps) => {
        const tickLabels: number[] = props.data[0].points.map((point: SeriesDataPoint) => point.y)
        const range: AxisRange = getYAxisRange(props.data, props.config);
        const interval = range.max - range.min;

        return (
            <g className={styles.YAxis__Ticks}>
                {tickLabels.map((label: number) => {
                    const y1 = ((100 - props.padding) * (1 - (label - range.min) / interval)) + 10;
                    const textYPos = Math.max(Math.min(y1, (100 - props.padding / 2)), props.padding / 2);
                    return (
                        <g>
                            <line x1={'90%'} y1={`${y1}%`} x2={'100%'} y2={`${y1}%`} stroke={'black'}/>
                            <text className={styles.YAxis__TickLabels} x={'60%'} y={`${textYPos}%`} stroke={'black'}>{label}</text>
                        </g>
                    )}
                )}
            </g>
        )
    }
    
    return (
        <div className={styles.YAxis}>
            <svg className={styles.YAxis__Svg}>
                <line x1={'100%'} y1={'10%'} x2={'100%'} y2={'100%'} stroke={'black'} fill={'transparent'} strokeWidth={1}/>
                {renderTicks({data: props.data, padding: props.padding.right, config: props.config})}
            </svg>
        </div>
    )
};

export default XAxis;
