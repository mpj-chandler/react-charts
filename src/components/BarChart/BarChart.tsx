import React from 'react';
import styles from './BarChart.scss';
import XAxis from '../XAxis/XAxis';
import YAxis from '../YAxis/YAxis';
import { SeriesData, Padding } from './types';

export interface BarChartProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
}

const BarChart: React.FC<BarChartProps> = (props: BarChartProps) => {
    return (
        <div className={styles.BarChart}>
            <div className={styles.BarChart__title}>{props.title}</div>
            <YAxis title={'YAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__blank}>()</div>
            <XAxis title={'XAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__padRight}>()</div>
        </div>
    )
}

export default BarChart;
