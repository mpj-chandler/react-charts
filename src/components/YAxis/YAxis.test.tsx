import React from 'react';
import TestRenderer from 'react-test-renderer';
import YAxis from './YAxis';
import { AxisProps, AxisConfig } from '../BarChart/types';
import Placement from '../../enums/Placement';


jest.mock('../AxisTickLabel/AxisTickLabel');

describe('YAxis', () => {
    const props: AxisProps = {
        title: 'anything',
        data: [
            {
                seriesName: 'A',
                points: [
                    {
                        x: 0,
                        y: 10
                    },
                    {
                        x: 1,
                        y: 20
                    },
                    {
                        x: 2,
                        y: 30
                    }
                ]
            }
        ],
        padding: {
            top: 10,
            bottom: 10,
            left: 10,
            right: 10,
        }
    };

    const config: AxisConfig = {
        zeroIntercept: true,
        margin: 20,
        tickPlacement: Placement.Aligned,
        tickLength: 10,
    }

    it('it renders consistently', () => {
        const component = TestRenderer.create(<YAxis {...props}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with an explicit config', () => {
        const component = TestRenderer.create(<YAxis {...props} config={{ }}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });
})