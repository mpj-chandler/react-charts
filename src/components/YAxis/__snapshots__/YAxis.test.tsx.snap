// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YAxis it renders consistently 1`] = `
<div
  className="YAxis"
>
  <svg
    className="YAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="10%"
      x2="10%"
      y1="10%"
      y2="90%"
    />
    <g
      className="YAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="90%"
          y2="90%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="90%"
        >
          10
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="50%"
          y2="50%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="50%"
        >
          20
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="10%"
          y2="10%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="10%"
        >
          30
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`YAxis it renders consistently with an explicit config 1`] = `
<div
  className="YAxis"
>
  <svg
    className="YAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="10%"
      x2="10%"
      y1="10%"
      y2="90%"
    />
    <g
      className="YAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="NaN%"
          x2="10%"
          y1="90%"
          y2="90%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="NaN%"
          y="90%"
        >
          10
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="NaN%"
          x2="10%"
          y1="50%"
          y2="50%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="NaN%"
          y="50%"
        >
          20
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="NaN%"
          x2="10%"
          y1="10%"
          y2="10%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="NaN%"
          y="10%"
        >
          30
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`YAxis it renders consistently without a config 1`] = `
<div
  className="YAxis"
>
  <svg
    className="YAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="10%"
      x2="10%"
      y1="10%"
      y2="90%"
    />
    <g
      className="YAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="90%"
          y2="90%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="90%"
        >
          10
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="50%"
          y2="50%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="50%"
        >
          20
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="10%"
          y2="10%"
        />
        <text
          className="YAxis__TickLabels"
          stroke="black"
          x="5%"
          y="10%"
        >
          30
        </text>
      </g>
    </g>
  </svg>
</div>
`;
